### DISCUSSION TITLE
LittleFS can't read last file created

### DISCUSSION THREAD

**Discussion by d-soliveira** [zephyrproject-rtos/zephyr#36775] [Q&A] [answered]
Hi there, 

I created an filesystem image on Ubuntu as described in [Littlefs-fuse](https://github.com/littlefs-project/littlefs-fuse), with a simple TXT file just for testing and flashed it at the correct uC's memory address.

For some reason, I cannot read the TXT file content. The function _fs_open(&file, fname_txt, FS_O_READ)_ returns 0, which means it found the file, but the _fs_read(&file, &buffer, sizeof(buffer))_ returns 0 and the buffer array is empty.

However, if I create a second file I can read the first file but not the second file and so on. The new file could be empty.

I guess it could be some image's configuration, but I can't figure it out. 
Someone have experience something similar?

Thank you!

**Comment by de-nordic** [comment, marked as answer]
Have you provided the lfs, when creating the image, with following parameters:
  `--block-size` equal to flash device erase page size
  `--read_size`
  `--prog_size`
  `--block_cycles`
  `--cache_size`
  `--lookahead_size`

according to your board DTS/Kconfig parameters?

**Reply by d-soliveira** [reply]
Hi @de-nordic !

I haven't configured the LFS parameters, probably that is the problem.
I'm using an STM32L4S5VI, which has 256 pages of 4 Kbytes and my DTS file is:
```
storage_partition: partition@fc000 {
			label = "storage";
			reg = <0x000fc000 0x0004000>;
```

I tried to format the LFS as you suggested, but now I'm getting another error when try to mount it on Linux host
```` 
littlefs/lfs.c:1070:error: Corrupted dir pair at {0x0, 0x1}
lfs_fuse.c:514:error: Invalid or incomplete multibyte or wide character
````

My steps were:
1.  Create an image `dd if=/dev/zero of=image bs=4096 count=4 `
2.  Format the LFS image `./lfs --block_size=4096 --block_count=4 --read_size=4096 --prog_size=4096  --format /dev/loop12`
3. Mount it on Ubuntu `./lfs /dev/loop12 mount` and got the error.

If I understand, the block_size is each small page's block (4 Kb in my case) and the block_count are the number of 4Kb I need. Am I right?

The LFS's documentation states the 

```
    // Size of an erasable block. This does not impact ram consumption and
    // may be larger than the physical erase size. However, non-inlined files
    // take up at minimum one block. Must be a multiple of the read
    // and program sizes.
    lfs_size_t block_size;

    // Number of erasable blocks on the device.
    lfs_size_t block_count;
```

Thank you

**Reply by de-nordic** [reply]
The default parameters are as follows:
![image](https://user-images.githubusercontent.com/56024351/124913343-4d2e7480-dfef-11eb-8abb-7bb46595d267.png)

as long as you do not use fstab, like the automount in samples/subsys/fs/littlefs does, these will be your defaults. The automount uses DTS overlay with the parameters.
The `read_size` and the `write_size` set granular size of the read/write operations within the block.
So the block size is fine (4k), the other need correction according to your setup.

**Reply by d-soliveira** [reply]
Hi @de-nordic 

I have configured the parameters as your suggestion, and now I can read the file without any problem. Thank you!

For someone who could be facing the same problem, my steps were:

1. Removed the fstab option.
2. Adjusted the DTS file to `storage_partition: partition@fc000 {
			label = "storage";
			reg = <0x000fc000 0x0008000>;
		};`
3. `dd if=/dev/zero of=image_new.bin bs=4096 count=8`
4. `./lfs --block_size=4096 --read_size=16 --prog_size=64 --block_cycles=512 --cache_size=64 --lookahead_size=32 --format /dev/loop12`
5. `./lfs --block_size=4096 --read_size=16 --prog_size=64 --block_cycles=512 --cache_size=64 --lookahead_size=32 /dev/loop12 mount`
